File: backend/python/src/data_go_kr/api_client.py
import requests
import xml.etree.ElementTree as ET
from typing import Dict, List, Optional, Any
from datetime import datetime, date
import json
import time
from urllib.parse import urlencode

from utils.utils_log import setup_logger

logger = setup_logger(__name__)


class DataGoKrApiClient:
    """공공데이터포털 조달청 나라장터 입찰공고정보서비스 API 클라이언트"""
    
    BASE_URL = "https://apis.data.go.kr/1230000/BidPublicInfoService04"
    
    def __init__(self, service_key: str, timeout: int = 30):
        """
        Args:
            service_key: 공공데이터포털에서 발급받은 서비스키
            timeout: API 요청 타임아웃 (초)
        """
        self.service_key = service_key
        self.timeout = timeout
        self.session = requests.Session()
    
    def _make_request(self, endpoint: str, params: Dict[str, Any]) -> requests.Response:
        """API 요청 실행"""
        url = f"{self.BASE_URL}/{endpoint}"
        
        # 기본 파라미터 추가
        params.update({
            'serviceKey': self.service_key,
            'type': 'xml'  # XML 응답 형식 지정
        })
        
        logger.info(f"API 요청: {url} | 파라미터: {params}")
        
        try:
            response = self.session.get(url, params=params, timeout=self.timeout)
            response.raise_for_status()
            
            logger.info(f"API 응답 성공: status_code={response.status_code}")
            return response
            
        except requests.exceptions.RequestException as e:
            logger.error(f"API 요청 실패: {str(e)}")
            raise
    
    def _parse_xml_response(self, response: requests.Response) -> Dict[str, Any]:
        """XML 응답 파싱"""
        try:
            root = ET.fromstring(response.content.decode('utf-8'))
            
            # 응답 헤더 정보 파싱
            header = root.find('.//header')
            if header is not None:
                result_code = header.find('resultCode')
                result_msg = header.find('resultMsg')
                
                if result_code is not None and result_code.text != '00':
                    error_msg = result_msg.text if result_msg is not None else '알 수 없는 오류'
                    logger.error(f"API 오류: {result_code.text} - {error_msg}")
                    raise Exception(f"API 오류: {result_code.text} - {error_msg}")
            
            # body 데이터 파싱
            items = []
            body = root.find('.//body')
            if body is not None:
                items_element = body.find('items')
                if items_element is not None:
                    for item in items_element.findall('item'):
                        item_data = {}
                        for child in item:
                            item_data[child.tag] = child.text
                        items.append(item_data)
                
                # 총 개수 정보
                total_count = body.find('totalCount')
                num_of_rows = body.find('numOfRows')
                page_no = body.find('pageNo')
                
                return {
                    'items': items,
                    'totalCount': int(total_count.text) if total_count is not None else 0,
                    'numOfRows': int(num_of_rows.text) if num_of_rows is not None else 0,
                    'pageNo': int(page_no.text) if page_no is not None else 1
                }
            
            return {'items': [], 'totalCount': 0, 'numOfRows': 0, 'pageNo': 1}
            
        except ET.ParseError as e:
            logger.error(f"XML 파싱 오류: {str(e)}")
            logger.error(f"응답 내용: {response.text[:1000]}")
            raise Exception(f"XML 파싱 오류: {str(e)}")
    
    def get_bid_list(
        self,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None,
        page_no: int = 1,
        num_of_rows: int = 100,
        area_code: Optional[str] = None,
        org_name: Optional[str] = None,
        bid_kind: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        입찰공고목록 정보 조회 (getBidPblancListInfoServc)
        
        Args:
            start_date: 검색 시작일 (기본값: 오늘)
            end_date: 검색 종료일 (기본값: 오늘)
            page_no: 페이지 번호 (기본값: 1)
            num_of_rows: 한 페이지 결과 수 (기본값: 100, 최대: 999)
            area_code: 지역코드
            org_name: 기관명
            bid_kind: 입찰종류 (01: 일반, 02: 제한, 03: 지명, 04: 턴키)
            
        Returns:
            입찰공고 목록 데이터
        """
        if start_date is None:
            start_date = date.today()
        if end_date is None:
            end_date = date.today()
        
        params = {
            'inqryBgnDt': start_date.strftime('%Y%m%d'),
            'inqryEndDt': end_date.strftime('%Y%m%d'),
            'pageNo': page_no,
            'numOfRows': min(num_of_rows, 999)  # 최대 999개로 제한
        }
        
        # 선택적 파라미터 추가
        if area_code:
            params['areaCd'] = area_code
        if org_name:
            params['instNm'] = org_name
        if bid_kind:
            params['bidKind'] = bid_kind
        
        response = self._make_request('getBidPblancListInfoServc', params)
        return self._parse_xml_response(response)
    
    def get_all_bid_list(
        self,
        start_date: Optional[date] = None,
        end_date: Optional[date] = None,
        area_code: Optional[str] = None,
        org_name: Optional[str] = None,
        bid_kind: Optional[str] = None,
        delay_seconds: float = 0.1
    ) -> List[Dict[str, Any]]:
        """
        전체 입찰공고목록 정보 조회 (페이지네이션 자동 처리)
        
        Args:
            start_date: 검색 시작일
            end_date: 검색 종료일
            area_code: 지역코드
            org_name: 기관명
            bid_kind: 입찰종류
            delay_seconds: 페이지 간 지연시간 (API 제한 고려)
            
        Returns:
            전체 입찰공고 목록
        """
        all_items = []
        page_no = 1
        num_of_rows = 999  # 최대값 사용
        
        logger.info(f"전체 입찰공고 조회 시작: {start_date} ~ {end_date}")
        
        while True:
            try:
                result = self.get_bid_list(
                    start_date=start_date,
                    end_date=end_date,
                    page_no=page_no,
                    num_of_rows=num_of_rows,
                    area_code=area_code,
                    org_name=org_name,
                    bid_kind=bid_kind
                )
                
                items = result.get('items', [])
                if not items:
                    break
                
                all_items.extend(items)
                logger.info(f"페이지 {page_no} 조회 완료: {len(items)}건 (누적: {len(all_items)}건)")
                
                # 마지막 페이지 확인
                if len(items) < num_of_rows:
                    break
                
                page_no += 1
                
                # API 제한을 고려한 지연
                if delay_seconds > 0:
                    time.sleep(delay_seconds)
                    
            except Exception as e:
                logger.error(f"페이지 {page_no} 조회 중 오류: {str(e)}")
                break
        
        logger.info(f"전체 입찰공고 조회 완료: 총 {len(all_items)}건")
        return all_items


def test_api():
    """API 테스트 함수"""
    # 테스트용 서비스키 (실제 사용시 환경변수에서 가져와야 함)
    service_key = "YOUR_SERVICE_KEY_HERE"
    
    client = DataGoKrApiClient(service_key)
    
    try:
        # 오늘 날짜로 테스트
        result = client.get_bid_list(num_of_rows=10)
        print(f"총 건수: {result['totalCount']}")
        print(f"조회된 건수: {len(result['items'])}")
        
        # 첫 번째 항목 출력
        if result['items']:
            item = result['items'][0]
            print("\n=== 첫 번째 입찰공고 ===")
            for key, value in item.items():
                print(f"{key}: {value}")
                
    except Exception as e:
        print(f"테스트 실패: {str(e)}")


if __name__ == "__main__":
    test_api()

File: backend/nodejs/api/nara/src/api-client.ts
/**
 * Korean Government Public Data Portal (data.go.kr) API client
 * Fetches bid notice information from the public procurement system
 */

import axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
import { parseStringPromise } from 'xml2js';

export interface ApiRequestParams {
  pageNo?: number;
  numOfRows?: number;
  inqryDiv?: string;
  inqryBgnDt?: string;
  inqryEndDt?: string;
  type?: 'xml' | 'json';
}

export interface ApiResponse<T = any> {
  header: {
    resultCode: string;
    resultMsg: string;
  };
  body: {
    pageNo: number;
    numOfRows: number;
    totalCount: number;
    items: T;
  };
}

export interface BidNoticeRawItem {
  bidNtceNo?: string;           // 입찰공고번호
  bidNtceNm?: string;           // 입찰공고명
  ntceInsttNm?: string;         // 공고기관명
  dminsttNm?: string;           // 수요기관명
  ntceInsttCd?: string;         // 공고기관코드
  dminsttCd?: string;           // 수요기관코드
  bidNtceDt?: string;           // 입찰공고일시
  bidNtceStDt?: string;         // 입찰공고시작일시
  bidNtceEndDt?: string;        // 입찰공고종료일시
  bidClseDt?: string;           // 입찰마감일시
  opengDt?: string;             // 개찰일시
  presmptPrce?: string;         // 예정가격
  asignBdgtAmt?: string;        // 배정예산금액
  bidMethdCd?: string;          // 입찰방식코드
  bidMethdNm?: string;          // 입찰방식명
  cntrctCnclsMthdCd?: string;   // 계약체결방법코드
  cntrctCnclsMthdNm?: string;   // 계약체결방법명
  rcptRegistNo?: string;        // 접수등록번호
  rcptRegistNm?: string;        // 접수등록명
  indstrytyLmttClCd?: string;   // 업종제한구분코드
  indstrytyLmttClNm?: string;   // 업종제한구분명
  prtcptCnditnCd?: string;      // 참가조건코드
  prtcptCnditnNm?: string;      // 참가조건명
  prtcptCndtnCn?: string;       // 참가조건내용
  bidNtceDtlCn?: string;        // 입찰공고상세내용
  cntrctPrd?: string;           // 계약기간
  dlvryPlce?: string;           // 납품장소
  ntceKindNm?: string;          // 공고구분명
  [key: string]: any;           // Additional dynamic fields
}

export class DataGoKrApiClient {
  private readonly baseUrl = 'https://apis.data.go.kr/1230000/ad/BidPublicInfoService';
  private readonly httpClient: AxiosInstance;
  private readonly serviceKey: string;

  constructor(serviceKey: string) {
    this.serviceKey = serviceKey;
    this.httpClient = axios.create({
      timeout: 30000,
      headers: {
        'User-Agent': 'Mozilla/5.0 (compatible; BidNoticeBot/1.0)',
        'Accept': 'application/json, application/xml'
      }
    });

    // Add request/response interceptors for logging
    this.httpClient.interceptors.request.use(
      (config) => {
        console.log(`[API Request] ${config.method?.toUpperCase()} ${config.url}`);
        return config;
      },
      (error) => {
        console.error('[API Request Error]', error.message);
        return Promise.reject(error);
      }
    );

    this.httpClient.interceptors.response.use(
      (response) => {
        console.log(`[API Response] ${response.status} ${response.statusText} - ${response.config.url}`);
        return response;
      },
      (error) => {
        console.error(`[API Response Error] ${error.response?.status} ${error.response?.statusText} - ${error.config?.url}`);
        return Promise.reject(error);
      }
    );
  }

  /**
   * Get bid announcement list
   * 입찰공고 목록 조회
   */
  async getBidPblancListInfoServc(params: ApiRequestParams = {}): Promise<BidNoticeRawItem[]> {
    const endpoint = `${this.baseUrl}/getBidPblancListInfoServc`;

    const requestParams = {
      serviceKey: this.serviceKey,
      pageNo: params.pageNo || 1,
      numOfRows: params.numOfRows || 100,
      type: params.type || 'json',
      ...params
    };

    try {
      const response: AxiosResponse = await this.httpClient.get(endpoint, {
        params: requestParams,
        timeout: 30000
      });

      if (response.status !== 200) {
        throw new Error(`HTTP ${response.status}: ${response.statusText}`);
      }

      // Handle different response formats
      let parsedData: any;

      if (params.type === 'xml' || response.headers['content-type']?.includes('xml')) {
        // Parse XML response
        parsedData = await this.parseXmlResponse(response.data);
      } else {
        // Handle JSON response
        parsedData = response.data;
      }

      // Extract items from response structure
      const items = this.extractItemsFromResponse(parsedData);

      console.log(`[API] Retrieved ${items.length} bid notices`);
      return items;

    } catch (error: any) {
      console.error(`[API Error] ${error.message}`);

      if (error.response) {
        console.error(`Response Status: ${error.response.status}`);
        console.error(`Response Data: ${JSON.stringify(error.response.data, null, 2)}`);
      }

      throw new Error(`API request failed: ${error.message}`);
    }
  }

  /**
   * Get all bid notices with pagination
   * 모든 입찰공고를 페이징으로 조회
   */
  async getAllBidList(
    startDate?: Date,
    endDate?: Date,
    areaCode?: string,
    orgName?: string,
    bidKind?: string,
    maxPages: number = 10
  ): Promise<BidNoticeRawItem[]> {
    const allItems: BidNoticeRawItem[] = [];
    let currentPage = 1;
    let hasMore = true;

    // Format dates for API
    const formatDate = (date: Date): string => {
      return date.toISOString().slice(0, 10).replace(/-/g, '') + '0000'; // YYYYMMDD0000
    };

    const baseParams: ApiRequestParams = {
      numOfRows: 100,
      type: 'json'
    };

    // Add date filters if provided
    if (startDate) {
      baseParams.inqryBgnDt = formatDate(startDate);
    }
    if (endDate) {
      baseParams.inqryEndDt = formatDate(endDate);
    }

    console.log(`[API] Starting paginated fetch - Start: ${startDate?.toISOString()}, End: ${endDate?.toISOString()}`);

    while (hasMore && currentPage <= maxPages) {
      try {
        const params = {
          ...baseParams,
          pageNo: currentPage
        };

        console.log(`[API] Fetching page ${currentPage}...`);
        const pageItems = await this.getBidPblancListInfoServc(params);

        if (pageItems.length === 0) {
          console.log(`[API] No more items on page ${currentPage}, stopping`);
          hasMore = false;
          break;
        }

        allItems.push(...pageItems);
        console.log(`[API] Page ${currentPage}: ${pageItems.length} items (Total: ${allItems.length})`);

        // Check if this is the last page
        if (pageItems.length < (baseParams.numOfRows || 100)) {
          console.log(`[API] Received fewer items than requested, assuming last page`);
          hasMore = false;
        }

        currentPage++;

        // Add delay between requests to avoid rate limiting
        await this.delay(500);

      } catch (error: any) {
        console.error(`[API] Error fetching page ${currentPage}:`, error.message);
        hasMore = false;
      }
    }

    console.log(`[API] Completed paginated fetch: ${allItems.length} total items`);
    return allItems;
  }

  /**
   * Parse XML response to JSON
   */
  private async parseXmlResponse(xmlData: string): Promise<any> {
    try {
      const result = await parseStringPromise(xmlData, {
        explicitArray: false,
        ignoreAttrs: true,
        normalize: true,
        normalizeTags: true,
        trim: true
      });
      return result;
    } catch (error: any) {
      throw new Error(`XML parsing failed: ${error.message}`);
    }
  }

  /**
   * Extract items array from API response structure
   */
  private extractItemsFromResponse(data: any): BidNoticeRawItem[] {
    try {
      // Handle different response structures
      let items: any = null;

      // Try different possible paths for items
      if (data?.response?.body?.items?.item) {
        items = data.response.body.items.item;
      } else if (data?.body?.items?.item) {
        items = data.body.items.item;
      } else if (data?.items?.item) {
        items = data.items.item;
      } else if (data?.response?.body?.items) {
        items = data.response.body.items;
      } else if (Array.isArray(data)) {
        items = data;
      }

      if (!items) {
        console.warn('[API] No items found in response structure');
        return [];
      }

      // Ensure items is an array
      if (!Array.isArray(items)) {
        items = [items];
      }

      return items.filter(item => item && typeof item === 'object');

    } catch (error: any) {
      console.error('[API] Error extracting items from response:', error.message);
      return [];
    }
  }

  /**
   * Add delay between API calls
   */
  private delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  /**
   * Validate API response
   */
  private validateResponse(data: any): boolean {
    // Check for error responses
    if (data?.header?.resultCode && data.header.resultCode !== '00') {
      throw new Error(`API Error: ${data.header.resultMsg || 'Unknown error'}`);
    }

    if (data?.response?.header?.resultCode && data.response.header.resultCode !== '00') {
      throw new Error(`API Error: ${data.response.header.resultMsg || 'Unknown error'}`);
    }

    return true;
  }
}